trino/trino.go

```diff
@@ -932,48 +971,185 @@ func (qr *driverRows) fetch(allowEOF bool) error {
 		qr.columns[i] = col.Name
-		qr.coltype[i] = newTypeConverter(col.Type)
+		qr.coltype[i] = newTypeConverter(col.Type, col.TypeSignature)
 	}
 }
 
 type typeConverter struct {
-	typeName   string
-	parsedType []string // e.g. array, array, varchar, for [][]string
+	typeName      string
+	parsedType    []string
+	scanType      reflect.Type
+	typeSignature typeSignature
+	hasPrecision  bool
+	precision     int64
+	scale         int64
+	hasSize       bool
+	size          int64
+}
+
+func newTypeConverter(typeName string, signature typeSignature) *typeConverter {
+	result := &typeConverter{
+		typeName:      typeName,
+		parsedType:    getNestedTypes([]string{}, signature),
+		typeSignature: signature,
+	}
+	result.scanType = getScanType(result.parsedType)
+	switch signature.RawType {
+	case "char", "varchar":
+		if len(signature.Arguments) > 0 && signature.Arguments[0].Kind == "LONG" {
+			result.hasSize = true
+			result.size = signature.Arguments[0].long
+		}
+	case "decimal":
+		if len(signature.Arguments) > 1 {
+			result.hasPrecision = true
+			if signature.Arguments[0].Kind == "LONG" {
+				result.precision = signature.Arguments[0].long
+			}
+			if signature.Arguments[1].Kind == "LONG" {
+				result.scale = signature.Arguments[1].long
+			}
+		}
+	}
+	return result
 }
 
-func newTypeConverter(typeName string) *typeConverter {
-	return &typeConverter{
-		typeName:   typeName,
-		parsedType: parseType(typeName),
+func getNestedTypes(types []string, signature typeSignature) []string {
+	types = append(types, signature.RawType)
+	if len(signature.Arguments) > 0 {
+		switch signature.Arguments[0].Kind {
+		case KIND_TYPE:
+			types = getNestedTypes(types, signature.Arguments[0].typeSignature)
+		case KIND_NAMED_TYPE:
+			types = getNestedTypes(types, signature.Arguments[0].namedTypeSignature.TypeSignature)
+		}
 	}
+	return types
 }
 
-// parses Trino types, e.g. array(varchar(10)) to "array", "varchar"
-// TODO: Use queryColumn.TypeSignature instead.
-func parseType(name string) []string {
-	parts := strings.Split(strings.ToLower(name), "(")
-	if len(parts) == 1 {
-		return parts
-	}
-	last := len(parts) - 1
-	parts[last] = strings.TrimRight(parts[last], ")")
-	if len(parts[last]) > 0 {
-		if _, err := strconv.Atoi(parts[last]); err == nil {
-			parts = parts[:last]
+func getScanType(typeNames []string) reflect.Type {
+	var v interface{}
+	switch typeNames[0] {
+	case "boolean":
+		v = sql.NullBool{}
+	case "json", "char", "varchar", "varbinary", "interval year to month", "interval day to second", "decimal", "ipaddress", "unknown":
+		v = sql.NullString{}
+	case "tinyint", "smallint":
+		v = sql.NullInt32{}
+	case "integer":
+		v = sql.NullInt32{}
+	case "bigint":
+		v = sql.NullInt64{}
+	case "real", "double":
+		v = sql.NullFloat64{}
+	case "date", "time", "time with time zone", "timestamp", "timestamp with time zone":
+		v = sql.NullTime{}
+	case "map":
+		v = NullMap{}
+	case "array":
+		if len(typeNames) <= 1 {
```

Comment:
what exactly does this mean?

Reply:
This is to avoid reading unset array indices. I'll change this into an assert.